packages/engine/src/workerpool/mod.rs

```diff
@@ -200,8 +219,28 @@ impl WorkerPoolController {
                     self.send_to_worker(worker.index(), WorkerPoolToWorkerMsg::task(sim_id, task))
                 })?;
             }
-            EngineToWorkerPoolMsgPayload::Sync(sync_msg) => {
-                self.send_to_all_workers(WorkerPoolToWorkerMsg::sync(sim_id, sync_msg))?
+            EngineToWorkerPoolMsgPayload::Sync(sync) => {
+                let sync = if let SyncPayload::State(sync) = sync {
+                    sync
+                } else {
+                    self.send_to_all_workers(WorkerPoolToWorkerMsg::sync(sim_id, sync))?;
+                    return Ok(());
+                };
+
+                // TODO: Only send to workers that simulation run is registered with
+                let (worker_msgs, worker_completion_receivers) =
+                    sync.create_children(self.comms.num_workers());
+                for (worker_index, msg) in worker_msgs.into_iter().enumerate() {
+                    self.comms.send(worker_index, WorkerPoolToWorkerMsg {
+                        sim_id: Some(sim_id),
+                        payload: WorkerPoolToWorkerMsgPayload::Sync(SyncPayload::State(msg)),
+                    })?;
+                }
+                let fut = async move {
+                    let sync = sync; // Capture `sync` in lambda.
+                    sync.forward_children(worker_completion_receivers).await
+                };
+                pending_syncs.push(Box::pin(fut) as _);
```

Comment:
```suggestion
            EngineToWorkerPoolMsgPayload::Sync(sync_payload) => {
                if let SyncPayload::State(sync) = sync_payload {
                    // TODO: Only send to workers that simulation run is registered with
                    let (worker_msgs, worker_completion_receivers) =
                        sync.create_children(self.comms.num_workers());
                    for (worker_index, msg) in worker_msgs.into_iter().enumerate() {
                        self.comms.send(worker_index, WorkerPoolToWorkerMsg {
                            sim_id: Some(sim_id),
                            payload: WorkerPoolToWorkerMsgPayload::Sync(SyncPayload::State(msg)),
                        })?;
                    }
                    let fut = async move {
                        let sync = sync; // Capture `sync` in lambda.
                        sync.forward_children(worker_completion_receivers).await
                    };
                    pending_syncs.push(Box::pin(fut) as _);
                } else {
                    self.send_to_all_workers(WorkerPoolToWorkerMsg::sync(sim_id, sync))?;
                };
```

Does this work?

Reply:
I thought the first version was better, because the else branch is easier to notice and there's less nesting.

Reply:
Yeah fair, matter of taste. I think the `if let` statement looks pretty gross, especially with the 5 `syncs` in 3 lines. 
Happy to merge as is anyway though